File: internal/tools/tasks.go
package tools

import (
	"context"

	"github.com/modelcontextprotocol/go-sdk/mcp"

	"github.com/zamedic/labradoc-mcp/internal/labradoc"
)

// TasksCloseArgs represents input args for tasks_close tool.
type TasksCloseArgs struct {
	IDs []string `json:"ids" jsonschema:"Array of task IDs to close"`
}

// TasksCloseResult represents output for tasks_close tool.
type TasksCloseResult struct {
	Message string `json:"message"`
}

// TasksListResult represents output for tasks_list tool.
type TasksListResult = labradoc.TasksResponse

// TasksList lists tasks extracted from documents.
func TasksList(ctx context.Context, client *labradoc.Client) (*mcp.CallToolResult, TasksListResult, error) {
	result, err := client.TasksList(ctx)
	if err != nil {
		return errorResult(err), labradoc.TasksResponse{}, nil
	}

	return nil, *result, nil
}

// TasksClose closes/completes tasks by their IDs.
func TasksClose(ctx context.Context, client *labradoc.Client, args TasksCloseArgs) (*mcp.CallToolResult, TasksCloseResult, error) {
	if len(args.IDs) == 0 {
		return errorResult(errMissingRequired("ids")), TasksCloseResult{}, nil
	}

	err := client.TasksClose(ctx, args.IDs)
	if err != nil {
		return errorResult(err), TasksCloseResult{}, nil
	}

	return nil, TasksCloseResult{Message: "Tasks closed successfully"}, nil
}

File: src/tools/tasks.ts
import type { Client } from "../client.js";

export interface TasksCloseArgs {
  ids: string[];
}

export interface TasksCloseResult {
  message: string;
}

export type TasksListResult = Client extends { tasksList(): Promise<infer R> } ? R : never;

export function newErrorResult(msg: string) {
  return {
    content: [{ type: "text" as const, text: `Error: ${msg}` }],
    isError: true,
  };
}

export async function tasksList(client: Client) {
  try {
    const result = await client.tasksList();
    return {
      content: [{ type: "json" as const, text: JSON.stringify(result) }],
    };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}

export async function tasksClose(client: Client, args: TasksCloseArgs) {
  if (!args.ids?.length) return newErrorResult("missing required parameter: ids");
  try {
    await client.tasksClose(args.ids);
    return {
      content: [{ type: "text" as const, text: JSON.stringify({ message: "Tasks closed successfully" }) }],
    };
  } catch (e) {
    return newErrorResult(e instanceof Error ? e.message : String(e));
  }
}
